### PR TITLE
refactor: convert FlattenNode from a recursive algorithm to an iterative one

### PR DESCRIPTION
**Pull request by senojj**
#### What problem is being solved?
Previously, the function `FlattenNode`, in package `internal/modelgraph`, could trigger a stack overflow given a model of a sufficient size and shape. Below is a table summarizing the characteristics of the previous implementation:
```
┌──────────────────────────────────────┬───────────────────────────────────────────┐
│               Property               │                   Value                   │
├──────────────────────────────────────┼───────────────────────────────────────────┤
│ Recursion site                       │ model.go:108 — FlattenNode self-recursion │
├──────────────────────────────────────┼───────────────────────────────────────────┤
│ Attack chain (per relation)          │ define rN: rN-1 (pure computed chain)     │
├──────────────────────────────────────┼───────────────────────────────────────────┤
│ Proto bytes / level                  │ ~35.3 (measured @ depth 200k)             │
├──────────────────────────────────────┼───────────────────────────────────────────┤
│ Overflow depth (default 1 GiB stack) │ ~3,300,000                                │
├──────────────────────────────────────┼───────────────────────────────────────────┤
│ Min. proto model size to overflow    │ ≈ 125,000,000 bytes (~125 MB / 120 MiB)   │
└──────────────────────────────────────┴───────────────────────────────────────────┘
```
This can be a _real_ concern, given that the default size of a PostgreSQL `BYTEA` data type is 1 GB, and the default size of a MySQL `LONGBLOB` is 4 GB. However OpenFGA _does_ provide the configuration `MaxAuthorizationModelSizeInBytes` with a sane default value of 256 KiB for the protobuf serialized model.

However, any OpenFGA instance configured to allow a model of sufficient size (>= 120 MiB) would be pathological.

#### How is it being solved?
The recursive behavior of the algorithm is changed to be iterative.

#### What changes are made to solve it?
The function is refactored to iterate over a heap allocated stack that collects subsequent work on each iteration.

### Performance
The previous, recursive implementation's benchmarks are as follows:
```
goos: darwin
goarch: arm64
pkg: github.com/openfga/openfga/internal/modelgraph
cpu: Apple M3 Max
BenchmarkFlattenNode/computed_union-16         	 5802746	       197.1 ns/op	      80 B/op	       7 allocs/op
BenchmarkFlattenNode/top_level_union-16        	 7166674	       165.9 ns/op	      72 B/op	       5 allocs/op
BenchmarkFlattenNode/ttu_logical-16            	13400839	        89.64 ns/op	      40 B/op	       3 allocs/op
BenchmarkFlattenNode/nested_ttu_computed-16    	 5241954	       221.2 ns/op	     120 B/op	       7 allocs/op
BenchmarkFlattenNode/userset_union-16          	 9780582	       125.3 ns/op	      48 B/op	       4 allocs/op
BenchmarkFlattenNode/union_with_intersection-16         	12214672	        95.04 ns/op	      40 B/op	       3 allocs/op
BenchmarkFlattenNode/union_with_exclusion-16            	12092679	        99.42 ns/op	      40 B/op	       3 allocs/op
BenchmarkFlattenNode/wildcard_union-16                  	 8844222	       137.7 ns/op	      48 B/op	       4 allocs/op
BenchmarkFlattenNode/recursive_userset_skipped-16       	 7015111	       171.0 ns/op	      64 B/op	       5 allocs/op
BenchmarkFlattenNode/recursive_ttu_skipped-16           	 8060342	       149.2 ns/op	      56 B/op	       4 allocs/op
BenchmarkFlattenNode/recursive_selective_skip-16        	 2831368	       420.9 ns/op	     216 B/op	      13 allocs/op
BenchmarkFlattenNode/mixed_broad-16                     	 3998406	       302.1 ns/op	     184 B/op	       8 allocs/op
BenchmarkFlattenNode/pathological_deep_computed_chain-16         	  338116	      3491 ns/op	     808 B/op	     101 allocs/op
BenchmarkFlattenNode/pathological_wide_union-16                  	  118140	     10242 ns/op	    5192 B/op	     203 allocs/op
```

The new, iterative implementation's benchmarks are as follows:
```
goos: darwin
goarch: arm64
pkg: github.com/openfga/openfga/internal/modelgraph
cpu: Apple M3 Max
BenchmarkFlattenNode/computed_union-16         	 7642923	       148.5 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/top_level_union-16        	 6592215	       168.4 ns/op	      72 B/op	       3 allocs/op
BenchmarkFlattenNode/ttu_logical-16            	13821214	        85.17 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/nested_ttu_computed-16    	 8051700	       147.7 ns/op	      40 B/op	       2 allocs/op
BenchmarkFlattenNode/userset_union-16          	11420436	       103.3 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/union_with_intersection-16         	13361565	        89.24 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/union_with_exclusion-16            	13488734	        87.27 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/wildcard_union-16                  	10497577	       113.2 ns/op	      32 B/op	       2 allocs/op
BenchmarkFlattenNode/recursive_userset_skipped-16       	 7617033	       155.7 ns/op	      64 B/op	       3 allocs/op
BenchmarkFlattenNode/recursive_ttu_skipped-16           	 8802837	       136.7 ns/op	      64 B/op	       3 allocs/op
BenchmarkFlattenNode/recursive_selective_skip-16        	 4575440	       254.8 ns/op	      72 B/op	       3 allocs/op
BenchmarkFlattenNode/mixed_broad-16                     	 5918486	       201.2 ns/op	     112 B/op	       3 allocs/op
BenchmarkFlattenNode/pathological_deep_computed_chain-16         	  462822	      2557 ns/op	       8 B/op	       1 allocs/op
BenchmarkFlattenNode/pathological_wide_union-16                  	  139347	      8506 ns/op	    8880 B/op	      14 allocs/op
```

The results show that the iterative approach significantly reduces allocations per operation, as well as overall latency per operation.

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved authorization model graph flattening reliability for deeply nested relationship structures.
  * Replaced recursive flatten traversal with an iterative approach to reduce stack-related risk and preserve traversal order.
  * Streamlined graph error handling to surface lookup failures more consistently.
* **Tests**
  * Added benchmarks covering worst-case flattening scenarios and key edge cases (unions, wildcard behavior, TTU resolution, and recursive relation skipping).

### CODE REVIEW

File: internal/modelgraph/model.go [hunk 1]
```diff
@@ -75,43 +75,47 @@ func (m *AuthorizationModelGraph) GetDirectEdgeFromNodeForUserType(objectRelatio
...
 
-		if canFlatten {
-			res, err := m.FlattenNode(edge.GetTo(), userType, hasWildcardRequest, recursivePath)
-			if err != nil {
-				return nil, err
+			if canFlatten {
+				stack = append(stack, edge.GetTo())
+			} else if !recursivePath || edge.GetRecursiveRelation() == "" {
+				result = append(result, edge)
 			}
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/modelgraph/model.go]: addressed by preserving the sort order of DFS

**Comment by yxshwanth** [comment]
Hi @senojj, nice work making `FlattenNode` stack-safe!

Wanted to flag a related PR for awareness since it touches the same function: #3225 fixes a correctness bug ([#3195](https://github.com/openfga/openfga/issues/3195)) where `FlattenNode`'s `recursivePath bool` parameter was skipping *every* recursive edge while unwinding a recursive relation, instead of only the specific one being unwound - causing experimental `weighted_graph_check` to incorrectly return `false` for models with two or more independently-recursive relations. It changes the signature from `recursivePath bool` to `skipRecursiveRelation string`.

Since both PRs touch the recursion-skipping logic in `FlattenNode`, there may be a small conflict/rebase depending on merge order. No action needed on your end right now, just wanted to make sure it's on your radar in case it's useful context for the iterative rewrite. Would appreciate a look at #3225 whenever you have a chance!

File: internal/modelgraph/benchmarks_test.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+		graph, err := modelgraph.New(model)
+		require.NoError(b, err)
+
+		node, ok := graph.GetNodeByID(bc.nodeID)
+		require.True(b, ok, "node %q not found", bc.nodeID)
+
+		// Sanity-check the case actually flattens (and doesn't error) before we
+		// benchmark it, so a broken model surfaces as a failure not a fast loop.
+		_, err = graph.FlattenNode(node, bc.userType, bc.hasWildcardRequest, bc.skipRecursiveRelation)
+		require.NoError(b, err)
```
[the remark this answers is not in the document]
Comment (senojj) [reply internal/modelgraph/benchmarks_test.go:325]: This is not necessary for benchmarks at this time.

**Review by justincoh — APPROVED** [review]

**Review by adriantam — APPROVED** [review]
